# PipeIT batch processing

This repository contains scripts for batch processing of raw .bam files using [PipeIT](https://github.com/ckynlab/PipeIT) pipeline. [PipeIT](https://github.com/ckynlab/PipeIT) is a somatic variant calling pipeline specific for Ion Torrent sequencing data.

## Batch processing on UBELIX

Run the following command with the parameters:
```
python3 PipeIT_batch_run_ubelix.py
  -b B        Folder with .bam data
  -i I        PipeIT image file
  -t T        Input bam folder
  -e E        Target panel bed file
  -x X        Ion Xpress Barcodes xlsx file with columns 'Sample Name', 'Normalize by', and 'Ion Xpress Barcode'
  -m M        Email to report when jobs are done (optional)
  -s S        snpEff jar file location
```
For example
```
python3 PipeIT_batch_run_ubelix.py -i /home/ubelix/dbmr/ko20g613/PipeIT_1.2.13.img -b /storage/research/dbmr_urology/Prostate_PDO /home/ubelix/dbmr/ko20g613/PipeIT_1.2.13.img -t /storage/research/dbmr_urology/Prostate_PDO/bam -e /storage/research/dbmr_urology/Prostate_PDO/WG_IAD127899.20170720.designed.bed -x /storage/research/dbmr_urology/Prostate_PDO/20200716_prostate_panel_sequencing.xlsx -s /home/ubelix/dbmr/ko20g613/snpEff/SnpSift.jar  -m e.ovchinnikova@gmail.com 
```
The output .vcf and .tsv files for each sample will be located in the folder `PipeIT/results/<sample name>`.

After the processing has finished, you can collect all .tsv results in one file by running the following command
```
python3 collect_PipeIT_results.py
  -i I        Input folder with PipeIT results
  -x X        Ion Xpress Barcodes xlsx file with columns columns 'Sample Name', 'Normalize by', and 'Ion Xpress Barcode'
  -o O        Output .tsv file, default stdout

```
For example
```
python3 collect_PipeIT_results.py -i PipeIT/results -x /storage/research/dbmr_urology/Prostate_PDO/20200716_prostate_panel_sequencing.xlsx -o PDO_mutation_results.tsv
```

## Process one file

To process one file, run the following command
```
singularity run -B <path/to/folders/that/need/to/be/mounted> <path/to/PipeIT_<version>.img> -t path/to/tumor.bam -n path/to/normal.bam -e path/to/region.bed  -o <output_directory>
```
For example, for prostate PDO raw data run on UBELIX cluster:
```
singularity run -B /storage/research/dbmr_urology/Prostate_PDO PipeIT_1.2.13.img -t /storage/research/dbmr_urology/Prostate_PDO/bam/IonXpress_003.bam -n /storage/research/dbmr_urology/Prostate_PDO/bam/IonXpress_001.bam -e /storage/research/dbmr_urology/Prostate_PDO/WG_IAD127899.20170720.designed.bed -o test_3
```
where
- `/storage/research/dbmr_urology/Prostate_PDO` is the folder with prostate PDO raw data
- `PipeIT_1.2.13.img` is PipeIT singularity image that can be downloaded from PipeIT laboratory's website: [http://oncogenomicslab.org/software-downloads](http://oncogenomicslab.org/software-downloads)
- `/storage/research/dbmr_urology/Prostate_PDO/bam/IonXpress_003.bam` is the path to the tumour raw .bam file
- `/storage/research/dbmr_urology/Prostate_PDO/bam/IonXpress_001.bam`is the path to the blood raw .bam file used for normalizing
- `/storage/research/dbmr_urology/Prostate_PDO/WG_IAD127899.20170720.designed.bed` is the path to the .bed file with the targeted gene regions
- `test_3` is the name of the output folder that will be generated by PipeIT

For more details and options, see [the PipeIT repository docu](https://github.com/ckynlab/PipeIT).

PipeIT generates several output files in the folder `PipeIT/results/<output_directory>`. The .vcf can be found under `PipeIT/results/<output_directory>/<output_directory>.PipeIT.vcf`. To extract mutation information from this file in .tsv format, install [SnpEff & SnpSift](https://pcingola.github.io/SnpEff/download/) and run the following commands
```
module load Java/11.0.2
java -jar SnpSift.jar extractFields -s "," PipeIT/results/<output_directory>/<output_directory>.PipeIT.vcf CHROM POS REF ALT ANN[*].GENE ANN[*].HGVS_P AF > output.tsv
```

To extract more information from .vcf, check out [documentation of SnpSift](https://pcingola.github.io/SnpEff/ss_extractfields/#example-1-extracting-chromosome-position-id-and-allele-frequency).
